libgalois/include/katana/PropertyGraph.h

```diff
@@ -842,6 +852,513 @@ CreateSymmetricGraph(PropertyGraph* pg);
+
+    KATANA_LOG_DEBUG_ASSERT(dests_.size() == pg->topology().num_edges());
+    KATANA_LOG_DEBUG_ASSERT(dests_.size() == orig_edge_ids_.size());
+  }
+
+  const EdgeTypeIndex* edge_type_index_;
+  AdjIndexArray adj_indices_;
+  EdgeDstArray dests_;
+  OrigEdgeIDMap orig_edge_ids_;
+};
+
+/// Provides both out-going and in-coming topology API
+/// Provides Edge Type Aware topology API
+class KATANA_EXPORT EdgeTypeAwareBiDirTopology : public EdgeTypeAwareTopology {
+  // Inheriting to reuse outgoing topology API
+  //
+  using OutTopoBase = EdgeTypeAwareTopology;
+
+  using OutTopoBase::EdgeTypeID;
+
+public:
+  static EdgeTypeAwareBiDirTopology Make(const PropertyGraph* pg) noexcept {
+    // EdgeTypeIndex edge_type_index = EdgeTypeIndex::Make(pg);
+    // OutTopoBase out_topo = OutTopoBase::MakeFromDefaultTopology(pg, &edge_type_index);
+    // EdgeTypeAwareTopology in_topo = EdgeTypeAwareTopology::MakeFromTransposeTopology(pg, &edge_type_index);
+    // return EdgeTypeAwareBiDirTopology{std::move(edge_type_index), std::move(out_topo), std::move(in_topo)};
+
+    auto edge_type_index =
+        std::make_unique<EdgeTypeIndex>(EdgeTypeIndex::Make(pg));
+    return EdgeTypeAwareBiDirTopology{pg, std::move(edge_type_index)};
+  }
+
+  EdgeTypeAwareBiDirTopology() = default;
+  EdgeTypeAwareBiDirTopology(EdgeTypeAwareBiDirTopology&&) = default;
+  EdgeTypeAwareBiDirTopology& operator=(EdgeTypeAwareBiDirTopology&&) = default;
+
+  EdgeTypeAwareBiDirTopology(const EdgeTypeAwareBiDirTopology&) = delete;
+  EdgeTypeAwareBiDirTopology& operator=(const EdgeTypeAwareBiDirTopology&) =
+      delete;
+
+  bool has_edge_type_id(const EdgeTypeID& edge_type) const noexcept {
+    return edge_type_index_->has_edge_type_id(edge_type);
+  }
+
+  auto distinct_edge_type_ids() const noexcept {
+    return edge_type_index_->distinct_edge_type_ids();
+  }
+
+  edges_range in_edges(Node N) const noexcept { return in_topo_.edges(N); }
+
+  edges_range in_edges(Node N, const EdgeTypeID& edge_type) const noexcept {
+    return in_topo_.edges(N, edge_type);
+  }
+
+  Node in_edge_dest(Edge edge_id) const noexcept {
+    return in_topo_.edge_dest(edge_id);
+  }
+
+  size_t in_degree(Node N) const noexcept { return in_topo_.degree(N); }
+
+  size_t in_degree(Node N, const EdgeTypeID& edge_type) const noexcept {
+    return in_topo_.degree(N, edge_type);
+  }
+
+  Edge original_edge_id_using_in_edge(Edge in_edge) const noexcept {
+    return in_topo_.original_edge_id(in_edge);
+  }
+
+  /// Returns an edge iterator to an edge with some node and key by
+  /// searching for the key via the node's outgoing or incoming edges.
+  /// If not found, returns nothing.
+  edges_range FindAllEdgesSingleType(Node src, Node dst) const {
+    if (OutTopoBase::degree(src) == 0 || in_topo_.degree(dst) == 0) {
+      return MakeStandardRange<edge_iterator>(Edge{0}, Edge{0});
+    }
+
+    return OutTopoBase::FindAllEdgesSingleType(src, dst);
```

Comment:
It would be good to choose b/w out and in-edges search. The problem with the latter is that we can need to an easy way to convert in_edges_range to edges_range. Please add a TODO.